Product/Main.cpp

```diff
@@ -0,0 +1,334 @@
+
+    // Turn off all LEDs
+    for (int i = 0; i < numLEDs; i += 1) {
+        hw.digitalWrite(LEDpins[i], LED_OFF);
+    }
+
+    // Should we pulse the fan?
+    bool pulseFan = alert == alertFanRPM;
+
+    // Flash the alert LEDs, sound the buzzer, and pulse the fan, until the user powers off.
+    while (1) {
+        hw.analogWrite(BUZZER_PIN, BUZZER_ON);
+        hw.digitalWrite(ERROR_LED_PIN, LED_ON);
+        for (int i = 0; leds[i] != -1; i += 1) {
+            hw.digitalWrite(leds[i], LED_ON);
+        }
+        if (pulseFan) fanController.setDutyCycle(fanDutyCycles[fanMedium]);
+        hw.delay(DELAY_500ms);
+
+        hw.analogWrite(BUZZER_PIN, BUZZER_OFF);
+        hw.digitalWrite(ERROR_LED_PIN, LED_OFF);
+        for (int i = 0; leds[i] != -1; i += 1) {
+            hw.digitalWrite(leds[i], LED_OFF);
+        }
+        if (pulseFan) fanController.setDutyCycle(fanDutyCycles[fanLow]);
+        hw.delay(DELAY_200ms);
+    }
+}
+
+/********************************************************************
+ * Fan
+ ********************************************************************/
+
+// Set the fan to the indicated speed, and update the fan indicator LEDs.
+void setFanSpeed(FanSpeed speed)
+{
+    fanController.setDutyCycle(fanDutyCycles[speed]);
+    
+    hw.digitalWrite(FAN_LOW_LED_PIN,  LED_ON);
+    hw.digitalWrite(FAN_MED_LED_PIN,  speed >  fanLow  ? LED_ON : LED_OFF);
+    hw.digitalWrite(FAN_HIGH_LED_PIN, speed == fanHigh ? LED_ON : LED_OFF);
+
+    currentFanSpeed = speed;
+
+    // disable fan RPM monitor for a few seconds, until the new fan speed stabilizes
+    dontCheckFanSpeedUntil = hw.millis() + DELAY_3sec;
+}
+
+// Call this periodically to check that the fan RPM is within the expected range for the current FanSpeed.
+void updateFan() {
+    const unsigned int fanRPM = fanController.getSpeed(); 
+    // Note: we call getSpeed() even if we're not going to use the result, because getSpeed() works better if you call it often.
+
+    // If fan RPM checking is temporarily disabled, then do nothing.
+    if (dontCheckFanSpeedUntil) {
+        if (dontCheckFanSpeedUntil > hw.millis()) return;
+        dontCheckFanSpeedUntil = 0;
+    }
+
+    // If the RPM is too low or too high compared to the expected value, raise an alert.
+    const unsigned int expectedRPM = expectedFanRPM[currentFanSpeed];
+    if ((fanRPM < (lowestOkFanRPM * expectedRPM)) || (fanRPM > (highestOkFanRPM * expectedRPM))) {
+        enterAlertState(alertFanRPM);
+    }
+}
+
+/********************************************************************
+ * Battery
+ ********************************************************************/
+
+ // Return battery fullness as a number between 0 (empty = 12 volts) and 100 (full = 24 volts).
+unsigned int readBatteryFullness()
+{
+    // Read the current battery voltage. To smoothen the random variations in the
+    // reading, we take several readings and calculate the average.
+    const int numReadings = 10;
```

Comment:
I would move this to a thing that runs every event loop and calculates every N event loops like fan

Reply:
As it turns out, there is already code in updateBattery() that does exactly that. I deleted the code you highlighted in readBatteryFullness because we don't need it.